### PR TITLE
feat: Implement ColorPicker layout API

### PR DESCRIPTION
**Pull request by RedJue**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [X] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     Implement ColorPicker `layout` API       |
| 🇨🇳 Chinese |     实现 ColorPicker `layout` API       |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at bc51eee</samp>

This pull request adds a new `layout` prop to the `ColorPicker` component, which allows users to choose between a horizontal or vertical layout for the color picker panel. It also updates the documentation, the demo, the style, and the test of the component to support the new feature. It fixes a minor bug in the `ColorPresets` component and refactors some code for readability and consistency.

### <samp>🤖 Generated by Copilot at bc51eee</samp>

*  Add a new `layout` prop to the `ColorPicker` component, which allows switching between vertical and horizontal modes of the popup panel ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR45), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR75), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR149), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-3abf1a34a643a7087dfbaf3039447778574de2b61e627f4c1218297fb961c70fL18-R69), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-83dba3026e7dc41ad64d3a2f4d484eeb085f0c3798dc2316fbb90d23520fca6eR17-R18), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-83dba3026e7dc41ad64d3a2f4d484eeb085f0c3798dc2316fbb90d23520fca6eR29), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0R75-R106), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0L81-R134), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0R179-R180), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-34b89f88ff38f571cda106c0e4980f0487a530c9c21623a0fd1695bb0b132a58R28))
*  Modify the class names of the root and popup elements of the `ColorPicker` component, to add the `rtl` and `horizontal` modifiers based on the `direction` and `layout` props ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fL98-R107), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fL153-R160))
*  Export the `ColorPickerLayout` type and the `interface` module from the `index.ts` file of the `ColorPicker` component, to make them available for other modules ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-937e920c7edf3c1b09cde8ed592d2ccc37d8f6eaee8074582eec32dd70c90850L3-R5), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR22))
*  Define helper functions `genRtlStyle` and `genHorizontalLayoutStyle` to generate the CSS object for the right-to-left and horizontal layout modes of the `ColorPicker` component, in the `index.ts` file of the `ColorPicker` style module ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0R29-R46), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0R75-R106))
*  Add a link and a description of the new `layout` prop to the English and Chinese documents of the `ColorPicker` component ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560R30), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560R51), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249R31), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249R52))
*  Add a new demo code and document for the layout feature of the `ColorPicker` component, with a segmented control to switch between the layout modes ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-98588cbb9a9317be09ba0d73441ac2451d05c30d2c56ffa8ba7c3c419f1fc122R1-R7), [link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-2b037ba97634e8ea57d88193b551d6415a40911d28cc2d7ffbe070f070cfcf89R1-R73))
*  Add a new test case to check the horizontal layout mode of the `ColorPicker` component ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-aa06460351e8630f305cf6509e8137984e0f167ab2560da22874033189234fb4R254-R260))
*  Fix a missing dependency in the `useMemo` hook of the `ColorPresets` component, which renders the preset colors ([link](https://github.com/ant-design/ant-design/pull/42761/files?diff=unified&w=0#diff-8d2a305db0c4945017603df77e97b2bf27dbde09aef2a8e2c2f7d10e67a3a822L77-R77))

### CODE REVIEW

File: components/color-picker/index.en-US.md [hunk 1]
```diff
@@ -47,6 +48,7 @@ Used when the user needs to customize the color selection.
 | disabled | Disable ColorPicker | boolean | - |
 | placement | Placement of popup | `top` \| `topLeft` \| `topRight` \| `bottom` \| `bottomLeft` \| `bottomRight` | `bottomLeft` |
 | arrow | Configuration for popup arrow | `boolean \| { pointAtCenter: boolean }` | `true` | - |
+| layout | Layout of popup. | `horizontal` \| `vertical` | `vertical` |
```
Comment (li-jia-nan) [inline components/color-picker/index.en-US.md:51]: 后面得加个版本号 5.6.0

Comment (li-jia-nan) [reply components/color-picker/index.en-US.md:51]: 这么晚了，快睡觉吧铁子

File: components/color-picker/demo/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import { ColorPicker, Segmented, Space } from 'antd';
+import type { ColorPickerLayout } from 'antd/es/color-picker';
```
Comment (MadCcc) [inline components/color-picker/demo/layout.tsx:2]: demo 尽量不要用路径导入了，直接用 `ColorPickerProps['layout']`

File: components/color-picker/demo/layout.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                '#EB2F964D',
+              ],
+            },
+            {
+              label: 'Recent',
+              colors: ['#F5222D', '#FA8C16', '#FADB14'],
+            },
+          ]}
+          allowClear
+          autoAdjustOverflow={false}
```
Comment (MadCcc) [inline components/color-picker/demo/layout.tsx:66]: 这个还是开着

Comment (RedJue) [reply components/color-picker/demo/layout.tsx:66]: 行

File: components/color-picker/style/index.ts [hunk 4]
```diff
@@ -26,6 +26,24 @@ export const genActiveStyle = (token: ColorPickerToken) => ({
...
+      },
+      [`${componentCls}-clear`]: {
+        '&::after': {
+          direction: 'ltr',
+        },
+      },
+    },
+  };
+};
+
```
Comment (MadCcc) [inline components/color-picker/style/index.ts:46]: 混了点其他 PR 的进来？

Comment (RedJue) [reply components/color-picker/style/index.ts:46]: 应该是之前 pr 到 master 了， feature 还没合，合了就好了

File: components/color-picker/demo/layout.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import { ColorPicker, Segmented, Space } from 'antd';
+import type { ColorPickerProps } from 'antd/es/color-picker';
+import React, { useState } from 'react';
+
+const Demo = () => {
+  const [layout, setLayout] = useState<ColorPickerProps['layout']>('vertical');
```
Comment (afc163) [inline components/color-picker/demo/layout.tsx:6]: 默认展示 horizontal 吧，这样一眼就能看到新的能力。

Comment (RedJue) [reply components/color-picker/demo/layout.tsx:6]: 好 我改一下

**Review by kiner-tang — APPROVED** [review]

File: components/color-picker/ColorPickerPanel.tsx [hunk 6]
```diff
@@ -15,30 +15,60 @@ interface ColorPickerPanelProps extends ColorPickerBaseProps {
...
       )}
+      {innerPanelBodyRender(panel)}
+    </>
+  );
+
+  const extraPanelRender = (panel: React.ReactNode) => (
+    <div className={colorPickerPanelPrefixCls}>
+      {layout === 'horizontal' && (allowClear || presets)
+        ? horizontalLayoutRender(panel)
+        : defaultLayoutRender(panel)}
```
Comment (afc163) [inline components/color-picker/ColorPickerPanel.tsx:71]: 是不是应该直接把这个 pickerPanelRender 放出来，总是有自定义需求，layout 枚举感觉也不够用。

https://github.com/ant-design/ant-design/issues/42749

**Pull request closed without merging**